File: src/Historical/Pascal/flyweight.pas
program FlyweightExample;
{$mode objfpc}{$H+}

uses
  SysUtils;

type
  TStyle = record
    Font: string;
    Size: Integer;
    Color: string;
  end;

var
  Styles: array[0..1] of TStyle;
  Keys: array[0..1] of string;
  Count: Integer = 0;

function StyleKey(const Font: string; Size: Integer; const Color: string): string;
begin
  Result := Font + '|' + IntToStr(Size) + '|' + Color;
end;

function GetStyle(const Font: string; Size: Integer; const Color: string): Integer;
var
  Key: string;
  I: Integer;
begin
  Key := StyleKey(Font, Size, Color);
  for I := 0 to Count - 1 do
    if Keys[I] = Key then Exit(I);
  Keys[Count] := Key;
  Styles[Count].Font := Font;
  Styles[Count].Size := Size;
  Styles[Count].Color := Color;
  Result := Count;
  Inc(Count);
end;

var
  Red1, Red2, Blue: Integer;
begin
  Red1 := GetStyle('Inter', 12, 'red');
  Red2 := GetStyle('Inter', 12, 'red');
  Blue := GetStyle('Inter', 12, 'blue');
  if Blue < 0 then Halt(1);
  WriteLn('styles=', Count, ';shared=', LowerCase(BoolToStr(Red1 = Red2, True)), ';text=ABC');
end.

File: src/Historical/Fortran/flyweight.f90
program flyweight
  implicit none

  type :: text_style
    character(len=5) :: font
    integer :: size
    character(len=4) :: color
    logical :: used
  end type text_style

  type(text_style) :: pool(2)
  integer :: red1, red2, blue, styles

  pool%used = .false.
  red1 = get_style('Inter', 12, 'red ')
  red2 = get_style('Inter', 12, 'red ')
  blue = get_style('Inter', 12, 'blue')

  if (pool(blue)%color /= 'blue') error stop 'blue style missing'
  styles = count(pool%used)

  write (*,'(A,I0,A,A,A)') 'styles=', styles, ';shared=', &
       merge('true ', 'false', red1 == red2), ';text=ABC'

contains

  integer function get_style(font, font_size, color) result(id)
    character(len=*), intent(in) :: font, color
    integer, intent(in) :: font_size
    integer :: i

    do i = 1, size(pool)
      if (pool(i)%used .and. pool(i)%font == font .and. &
          pool(i)%size == font_size .and. pool(i)%color == color) then
        id = i
        return
      end if
    end do

    do i = 1, size(pool)
      if (.not. pool(i)%used) then
        pool(i) = text_style(font, font_size, color, .true.)
        id = i
        return
      end if
    end do

    error stop 'style pool exhausted'
  end function get_style
end program flyweight
